### ISSUE TITLE
Estimating fees

### ISSUE THREAD

**Issue by AdamISZ** [JoinMarket-Org/joinmarket#308]
Based on the discussion here: https://www.reddit.com/r/joinmarket/comments/3rm656/is_making_transactions_jm_in_these_days_of/

I'm putting together a PR to, optionally, calculate a fee estimate for fast/default/slow at the point when all the maker utxos have been received and we build the tx. It's based on this: 

    #Estimate fee per choice of 'fast','medium','patient'
    #Assuming p2pkh:
    #out: 8+1+3+2+20=34, in: 1+32+4+1+1+~73+1+1+33=147, ver:4,seq:4, +2 (len in,out)
    #total ~= 34*len_out + 147*len_in + 10 (sig sizes vary slightly)

Obviously there are several issues. First, if the change amount that gets calculated from this is less than the dust limit, the attempt would have to be abandoned. Second, I guess it can't be applied to a sweep. Third, having it optionally fail like that I guess would not work for a tumbler (well, not directly). Anyway putting it together and testing, and then see if we can refine it.

**Comment by AdamISZ** [comment]
Also, as we discussed on reddit, estimatefee is problematic - first, the senders/takers are usually/often not running Core, second I dont' think blockr exposes estimatefee in the API. I am putting default per kb values into the config for 'fast', 'default', 'slow' and going from there.

**Comment by AdamISZ** [comment]
OK, so going through it: in sendpayment, select_utxos() is called at the start, based on amount: cj_amount + total_cj_fee + txfee. Obviously that makes this more difficult. But I'm wondering if that call can't be delayed. Specifically, after recvtxio from all makers, one can do the fee estimate and then calculate the above, replacing the txfee value, and then call select_utxos.

I'm having a think about it but not sure why not.

**Comment by chris-belcher** [comment]
In this plan, would the taker be paying for extra fees if a maker sends lots of small UTXOs that increase the ?

It's probably worth doing it anyway because a transaction taking too long to confirm is more annoying than paying more in fees. But there might be an attack where a maker sends lots of dust knowing that somebody else will pay for it to be combined into a large UTXO. So maybe have a sane default ceiling on the amount a taker will pay, rotating to another maker if the required fee is too high.

For estimatefee I would structure it that a new abstract method is added to BlockchainInterface which for Bitcoin Core can call estimatefee, for blockr calls some other website, for an eventual ElectrumBlockchainInterface it can call the electrum function.

**Comment by AdamISZ** [comment]
Yes, I also thought about the dust side. But isn't that problem already there, in as much as currently the taker specifies the overall fee without knowing how many inputs the makers are going to give. It seems that the maker contribution should be a per-input amount rather than an absolute number.

I agree about estimatefee, that was also my initial idea, the problem I had today is that the APIs I looked at, blockr and blockchain.info, don't seem to expose any fee estimation. If the large majority of takers are using blockr, then the rpc call part is not going to be the important bit.

I'm looking at someone's bug right now, but my first stab at it stopped at my previous comment above.

**Comment by adlai** [comment]
People already relying on a 3rd party service (blockr) could rely on another, such as http://www.cointape.com/api

**Comment by AdamISZ** [comment]
Cool, thanks.

**Comment by AdamISZ** [comment]
So, the above PR is, briefly, like this:

* abstract estimate_fee_per_kb method implemented in BitcoinCoreInterface and BlockrInterface. The latter using cointape's API call, although I'm not happy with the numbers it's spitting out (80 satoshis per byte for half an hour seems excessive?).

* initial estimate of txfee is based on 5000 satoshis per counterparty (yuck), adjustable with txfee option. Added some notes there so people know what it means. Then, an agressive choice of amount for calling select_utxos() - assume transaction fee will be 3 times whatever that figure is. This is the part I'm least happy about. But, it means that when we calculate a realistic txfee based on the created tx size, we are far less likely to find that we get a negative change amount.

* the main edit is in recv_txio: once transaction is assembled, use the formula above to get the approximate transaction size (testing seems to show it's usually pretty close, a few bytes out). Apply whatever is thrown out by estimate_fee_per_kb.

* For sweeps it's obviously problematic, and I abandoned any effort: this code just takes the value in options.txfee and multiplies it by the number of makers, and sticks with that value for the fee.

* I haven't looked at tumbler yet, the main concern I have is recovering gracefully in case the change amount ends up calculated as negative (as above). But obviously this needs to be done.

Testing - have tried it on regtest, testnet and (once so far) on mainnet. Things are working, but: note it's quite tricky to get regtest to spit out any fee estimate, also note cointape's fee estimate from its API is unrealistically large, which is a bit of an annoyance (since that is very likely to get used by makers, as it stands). Also very tricky of course is that proper testing means testing out the edge cases which are really important here, but hard to trigger.

I'm sure there are improvements, please let me know what you think.

**Comment by AdamISZ** [comment]
This looks like a good option?

```
~$ curl https://api.blockcypher.com/v1/btc/main
{
  "name": "BTC.main",
  "height": 386659,
  "hash": "00000000000000000dfe160d665bad53a08b5400eee6a3f248f67624e559c0f1",
  "time": "2015-12-04T09:57:33.520135268Z",
  "latest_url": "https://api.blockcypher.com/v1/btc/main/blocks/00000000000000000dfe160d665bad53a08b5400eee6a3f248f67624e559c0f1",
  "previous_hash": "0000000000000000003c95fa0dde7a8994d64424748db0b7851ef45088736ae3",
  "previous_url": "https://api.blockcypher.com/v1/btc/main/blocks/0000000000000000003c95fa0dde7a8994d64424748db0b7851ef45088736ae3",
  "peer_count": 1374,
  "unconfirmed_count": 43604,
  "high_fee_per_kb": 47689,
  "medium_fee_per_kb": 24201,
  "low_fee_per_kb": 24038,
  "last_fork_height": 386456,
  "last_fork_hash": "0000000000000000059a2e23eabddcfc8ba1048ec3b9903128ddb1753e914b94"
}
```

These per kB fees look realistic.

**Comment by AdamISZ** [comment]
Successful tests of (a) successful spend, 27K fee, using blockr, and receiving estimate from blockcypher like above : 4489a8cc933cb4e94915ead5b57b4aa707212c1f7b317187b500491e068c7887

(b) failed spend, not enough funds to cover (3*initial estimate of: 3 counterparties * 5000 options.txfee value) + total cj fee. Tried this a couple of times.

(c) succesful test with 'just enough' funds, using bitcoinrpc and correctly receiving fee estimate for number of blocks (estimatefee 3 -> 14985 satoshis), resulting in 19K fee.
```
[2015/12/05 15:53:42] rpc: estimatefee [3]
[2015/12/05 15:53:42] got estimated tx bytes: 1311
[2015/12/05 15:53:42] Based on initial guess: 15000, we estimated a fee of: 19645
```

Testnet tests are ongoing, seems to be a problem with testnet at the moment.

Would appreciate any ideas about how to deal with tumbler - i.e. what should happen if it turns out the aggressive 3*estimate used for select_utxos is not enough.

**Comment by AdamISZ** [comment]
Fixed up sweeps and rationalise fee estimation a bit (pulled the reusable code out into bitcoin/main.py and common.py).

Earlier I said: 

* For sweeps it's obviously problematic, and I abandoned any effort: this code just takes the value in options.txfee and multiplies it by the number of makers, and sticks with that value for the fee.

But it isn't necessary to give up so easily, and since sweeps are a kind of transaction that's often abnormally large, it's highly undesirable too. As a reminder, the problem with sweeps specifically is effectively a lack of a 'free variable'; for ordinary transactions we can reset the change amount once we know the "correct" transaction fee. With sweeps there is no change, so the coinjoin amount must be fixed at the point of choosing orders and issuing (template) transactions to the counterparties. For this reason the fee estimate must be driven by the (often large list of) utxos that we are going to use as input. This code estimates the number of inputs as (our list of inputs + 3*number of makers), assuming that 3 inputs per maker is reasonable. However, to be conservative, if the value set in `options.txfee` * number of makers is > than that estimate, we choose the larger estimate. This will mean that most of the time a reasonably intelligent fee estimate is chosen.

I have tried it a couple of times on testnet and once on mainnet and the debug logs look reasonably sensible. E.g. 99c6754818120ab8c8e7f3d5726062cbe6370a51f31765d8ad4b00e289bb9ba6 on testnet, this was with tx_fees = 1 in joinmarket.cfg, so produced a high fee estimate (119k sats), and https://blockchain.info/tx/13e0f6110c14a0b101ffab96341608c8382b683f92f44348bac7702b1a2f2ce3 on mainnet with tx_fees = 3, so producing a medium fee estimate (28k sats).

I'll add a test for a sweep to the test script `fee-estimation.py`.

**Comment by AdamISZ** [comment]
Added changes for tumbler. It's basically the same for both sweep and normal transaction (albeit there is some code copying, refactoring is desirable). The main extra point is: the approach of "select utxos for amount required + 3*estimated txfee" can fail, so it catches the exception "Not enough funds" (we should make a more specific exception type) and retries with amount required + 1*estimated fee, in the hope that the transaction will not fail and thus abort the tumble.

Tests OK on tumbler-test.py.

I'll leave this for a couple of days to see if people have comments.

**Comment by AdamISZ** [comment]
#346 Updated for the refactoring changes, now ready to merge into develop branch. Have tested regtest, tumbler and wallet tests. 

(wasn't able to squash it properly, sorry for extra commit).

**Comment by AdamISZ** [comment]
Another big round of testing passed, merged #346 to develop. Please reopen if you think this is (too!) flawed.
